src/atom.rs

```diff
@@ -0,0 +1,166 @@
+use std::cmp::{Ord, Ordering, PartialOrd};
+use std::str::from_utf8_unchecked;
+
+#[derive(Eq, PartialEq)]
+pub enum Atom<'a> {
+    /// A sequence of digits.
+    Number(u64),
+    /// A sequence of characters.
+    Chars(&'a str),
+    /// A sequence of underscores.
+    Underscore(usize),
+}
+
+impl<'a> PartialOrd<Atom<'a>> for Atom<'a> {
+    fn partial_cmp(&self, other: &Atom<'a>) -> Option<Ordering> {
```

Comment:
This should be written as:

```rust
impl PartialOrd for Atom<'_> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
```

The PartialOrd type parameter is defaulted to the right thing.